test/e2e/storage/drivers/csi-test/mock/service/service.go

```diff
@@ -0,0 +1,326 @@
+package service
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock/cache"
+
+	"github.com/golang/protobuf/ptypes"
+)
+
+const (
+	// Name is the name of the CSI plug-in.
+	Name = "io.kubernetes.storage.mock"
+
+	// VendorVersion is the version returned by GetPluginInfo.
+	VendorVersion = "0.3.0"
+
+	// TopologyKey simulates a per-node topology.
+	TopologyKey = Name + "/node"
+
+	// TopologyValue is the one, fixed node on which the driver runs.
+	TopologyValue = "some-mock-node"
+)
+
+// Manifest is the SP's manifest.
+var Manifest = map[string]string{
+	"url": "https://k8s.io/kubernetes/test/e2e/storage/drivers/csi-test/mock",
+}
+
+type CreateVolumeHook func(context.Context, *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error)
+type CreateSnapshotHook func(context.Context, *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error)
+type NodeStageHook func(context.Context, *csi.NodeStageVolumeRequest) (*csi.NodeStageVolumeResponse, error)
+type NodeUnstageHook func(context.Context, *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error)
+type NodeGetVolumeStatsHook func(context.Context, *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error)
```

Comment:
So far I have only converted those hooks that are actually needed. Without generics, adding hooks to a function is a bit of (tedious) cut-and-paste.

Should we aim for instrumenting all functions from the beginning or can that be added as needed? That would work because the tests and the driver are now in the same repo.

Reply:
As mentioned in https://github.com/kubernetes/kubernetes/pull/97069#discussion_r572213634, I've now pushed a revision that doesn't add any of these manually defined hooks.